keyring.go

```diff
@@ -310,19 +347,48 @@ func (d *DirectKeyFetcher) FetchKeys(
 
 func (d *DirectKeyFetcher) fetchKeysForServer(
 	ctx context.Context, serverName ServerName, requests map[PublicKeyRequest]Timestamp,
-) (map[PublicKeyRequest]ServerKeys, error) {
-	results, err := d.Client.LookupServerKeys(ctx, serverName, requests)
+) (map[PublicKeyRequest]PublicKeyLookupResult, error) {
+	serverKeys, err := d.Client.LookupServerKeys(ctx, serverName, requests)
 	if err != nil {
 		return nil, err
 	}
 
-	for req, keys := range results {
+	results := map[PublicKeyRequest]PublicKeyLookupResult{}
+	for req, keys := range serverKeys {
 		// Check that the keys are valid for the server.
 		checks, _, _ := CheckKeys(req.ServerName, time.Unix(0, 0), keys, nil)
 		if !checks.AllChecksOK {
 			return nil, fmt.Errorf("gomatrixserverlib: key response direct from %q failed checks", serverName)
 		}
+
+		mapServerKeysToPublicKeyLookupResult(keys, results)
```

Comment:
I'm confused, `mapServerKeysToPublicKeyLookupResult` seems to add multiple keys to the results map, while we didn't before? Does `mapServerKeysToPublicKeyLookupResult` add results under different keys?

Reply:
No, I think the keys of the results map are the same, it's just that mapServerKeysToPublicKeyLookupResult turns the `ServerKeys` structures into `PublicKeyLookupResults`.

Compare the implementation of `LookupServerKeys` (https://github.com/matrix-org/gomatrixserverlib/blob/de5aedb583c4a3c26074421b94e4f5c1ac488cf2/client.go#L229) with that of `mapServerKeysToPublicKeyLookupResult`. Note that in both cases the results map key is `PublicKeyRequest{serverName, keyID}` for the relevant server/key.

Reply:
Ok, so `mapServerKeysToPublicKeyLookupResult` will insert multiple keys into the map each time, which may clobber existing entries, but that's fine because... we're just undoing the mess that `LookupServerKeys` made? And we're just putting our fingers in our ears when it comes to handling same key appearing with different validity periods?

This is quite a confusing flow to be honest, so a comment would be good

Reply:
You're right, it was confusing. 8e95683 changes `LookupServerKeys` so that it just returns the list: effectively the map-building it was doing is now pulled up a level.

In terms of the same key appearing multiple times: yes it's a mess, but it's no more of a mess than before, and this PR is really just a starting point for clearing up some of this mess.